x-pack/plugins/maps/public/actions/map_actions.ts

```diff
@@ -0,0 +1,352 @@
+}
+
+export function updateMapSetting(
+  settingKey: string,
+  settingValue: string | boolean | number | object
+) {
+  return {
+    type: UPDATE_MAP_SETTING,
+    settingKey,
+    settingValue,
+  };
+}
+
+export function mapReady() {
+  return (dispatch: Dispatch, getState: () => MapStoreState) => {
+    dispatch({
+      type: MAP_READY,
+    });
+
+    getWaitingForMapReadyLayerListRaw(getState()).forEach((layerDescriptor) => {
+      dispatch<any>(addLayer(layerDescriptor));
+    });
+
+    dispatch({
+      type: CLEAR_WAITING_FOR_MAP_READY_LAYER_LIST,
+    });
+  };
+}
+
+export function mapDestroyed() {
+  return {
+    type: MAP_DESTROYED,
+  };
+}
+
+export function mapExtentChanged(newMapConstants: { zoom: number; extent: MapExtent }) {
+  return async (dispatch: Dispatch, getState: () => MapStoreState) => {
+    const state = getState();
+    const dataFilters = getDataFilters(state);
+    const { extent, zoom: newZoom } = newMapConstants;
+    const { buffer, zoom: currentZoom } = dataFilters;
+
+    if (extent) {
+      let doesBufferContainExtent = false;
+      if (buffer) {
+        const bufferGeometry = turf.bboxPolygon([
+          buffer.minLon,
+          buffer.minLat,
+          buffer.maxLon,
+          buffer.maxLat,
+        ]);
+        const extentGeometry = turf.bboxPolygon([
+          extent.minLon,
+          extent.minLat,
+          extent.maxLon,
+          extent.maxLat,
+        ]);
+
+        doesBufferContainExtent = turfBooleanContains(bufferGeometry, extentGeometry);
+      }
+
+      if (!doesBufferContainExtent || currentZoom !== newZoom) {
+        const scaleFactor = 0.5; // TODO put scale factor in store and fetch with selector
+        const width = extent.maxLon - extent.minLon;
+        const height = extent.maxLat - extent.minLat;
+        dataFilters.buffer = {
+          minLon: extent.minLon - width * scaleFactor,
+          minLat: extent.minLat - height * scaleFactor,
+          maxLon: extent.maxLon + width * scaleFactor,
+          maxLat: extent.maxLat + height * scaleFactor,
+        };
+      }
+    }
+
+    dispatch({
+      type: MAP_EXTENT_CHANGED,
+      mapState: {
+        ...dataFilters,
+        ...newMapConstants,
+      },
+    });
+    await dispatch<any>(syncDataForAllLayers());
+  };
+}
+
+export function setMouseCoordinates({ lat, lon }: { lat: number; lon: number }) {
+  let safeLon = lon;
+  if (lon > 180) {
+    const overlapWestOfDateLine = lon - 180;
+    safeLon = -180 + overlapWestOfDateLine;
+  } else if (lon < -180) {
+    const overlapEastOfDateLine = Math.abs(lon) - 180;
+    safeLon = 180 - overlapEastOfDateLine;
+  }
+
+  return {
+    type: SET_MOUSE_COORDINATES,
+    lat,
+    lon: safeLon,
+  };
+}
+
+export function clearMouseCoordinates() {
+  return { type: CLEAR_MOUSE_COORDINATES };
+}
+
+export function disableScrollZoom() {
+  return { type: SET_SCROLL_ZOOM, scrollZoom: false };
+}
+
+export function fitToLayerExtent(layerId: string) {
```

Comment:
Most other places `layerId` is set to type `string | null`, should that be the case here? Also, given how much we use `layerId`, would it be worth creating a type for it?

Reply:
It would not be accurate to create a layerId type that is `string | null`. A layerId is a `string`.

There are many action methods that take either `string | null`. `string` in the case that the layer is known and `null` when the layer is not known.  In the case of fitToLayerExtent action creation function, it only takes a string. The layer is known when  fitToLayerExtent is called so layerId will always be a string.